File: exo_stand_status/scripts/save_msgs.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# 订阅/joint_states/position消息，并保存为txt文件

import rospy
from sensor_msgs.msg import JointState

def joint_states_callback(msg):
    with open('/home/hhz/msgs/joint_states.txt', 'a') as file:
        position_str = ', '.join(str(val) for val in msg.position)
        file.write(position_str + '\n')

if __name__ == '__main__':
    rospy.init_node('joint_states_subscriber', anonymous=True)
    rospy.Subscriber('/joint_states', JointState, joint_states_callback)
    rospy.spin()

File: exo_stand_status/src/save_msgs.cpp
#include <ros/ros.h>
#include <sensor_msgs/JointState.h>
#include <fstream>

// 订阅/joint_states/position消息，并保存为txt文件

void jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg)
{
    std::ofstream file("/home/hhz/msgs/joint_states2.txt", std::ios_base::app);
    if (!file.is_open())
    {
        ROS_ERROR_STREAM("Could not open the file.");
        return;
    }

    for (size_t i = 0; i < msg->position.size(); ++i)
    {
        file << msg->position[i];
        if (i < msg->position.size() - 1)
            file << ", ";
    }
    file << std::endl;
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "joint_states_subscriber");
    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe("/joint_states", 10, jointStatesCallback);

    ros::spin();
    
    return 0;
}
